Added the possibility to configure multiple destinations

jbtrystram opened the pull request:
Using to the domain name.

The goal of this PR is two resolve the problem that you cannot have multiple container binded to the same port (obviously..)
You can now run the container with `-e DOMAIN="domain1.com;domain2.com"`
The UPSTREAM env value need to specify the different backend services, 1 per domain.

I also updated the FROM directory in the dockerfile, in order to pull the latest version of alpine.
Bash is also installed to help handle the arrays needed in the script.

Few changes have also been made in the letsencrypt command, which was outdated (the --authenticator option is no longer recognized, as well as --renew by default).
I tested this on my machine, it's building and running fine.
I hope this help, please let me know if you have any issues :)

jbtrystram committed:
Added the possibility to configure multiple destinations, according to the domain name.
You can now run the container with `-e DOMAIN="domain1.com;domain2.com"`
The UPSTREAM env value need to specify the different backend services, 1 per domain.

I also updated the FROM directory in the dockerfile, in order to pull the latest version of alpine.
Bash is also installed to halp handle the arrays needed in the script.

Few changes have also been made in the letsencrypt command, which was outdated (the --authenticator option is no longer recognized, as well as --renew by default).
I tested this on my machine, it's building and running fine.
I hope this help, please let me know if you have any issues :)

rgbkrk commented:
Thank you for diving in to update lets nginx!

If you break this up into smaller PRs it will be easier to handle reviewing and merging.

smashwilson commented on entrypoint.sh:
```diff
@@ -41,43 +57,55 @@ chown nginx:nginx /var/cache/nginx
+      --domain "${d}" \
+      --standalone \
+      --email "${EMAIL}" --agree-tos
+  fi
```
Rather than running `letsencrypt certonly` for each domain, it's better to run `letsencrypt certonly` a single time and pass multiple `--domain` arguments. That way you'll generate a single TLS certificate with many subject alternate names and not use up your five-certificates-in-seven-days quota.

jbtrystram replied:
This is a very good point! However, I'd say it's not a good idea, security-wise. Because if you share the same certificate between sercices and that certificate is compromised, then all the  services are compromised. Having different certs reduces the couplings between the served containers...

Oh, and I'm wondering on what the letsencrypt limitation is based. If it's based on the FQDN then you have plenty of room. If it's on the source IP then, indeed, we might hit the limit very quickly.

EDIT : according to [this](https://community.letsencrypt.org/t/rate-limits-for-lets-encrypt/6769), the limit is set for _given combination of Public Suffix + Domain (a "registered domain")_.
So that might be an issue if you pass multiple sub domains of the same domain([see here](https://community.letsencrypt.org/t/subdomains-certificates-are-limited-by-parent-domain/7148))
However, if you pass different domains (_e.g._ `domain1.com;domain2.com`) you are not affected by the limitation.

smashwilson commented on entrypoint.sh:
```diff
@@ -41,43 +57,55 @@ chown nginx:nginx /var/cache/nginx
+    --webroot \
+    -w /etc/letsencrypt/webrootauth/ \
+    --domain "${d}" \
+    --email "${EMAIL}" --agree-tos
```
Same here; this should still be a single script that runs `letsencrypt certonly --force-renewal` with many `--domain` arguments.

smashwilson commented on entrypoint.sh:
```diff
@@ -41,43 +57,55 @@ chown nginx:nginx /var/cache/nginx
+mkdir -p /etc/nginx/vhosts/
+
+# Copy nginx main conf file.
+cp templates/nginx.conf /etc/nginx/nginx.conf
```
Can we preserve the templating logic (`sed`, etc) for `nginx.conf` here, even though the default template doesn't use it? That way, if anyone has a customized `templates/nginx.conf` that's still a single file in a derived container image, it'll continue to work.

jbtrystram replied:
Yes, I've been wondering about that. Though you can still customize the way nginx globally behave by editing the nginx.conf file, you can't tune the behaviour on a vhost basis, since I chose to use the same template.

I think that we could look if there is a `templates/${DOMAIN}.conf` file that exist and use it if so, and if not use the default file. 
That would allow doing some tuning on a vhost/domain basis, the drawback being that you loose some automation (but if you need specific options on a specific vhost you won't have automation anyway, right?)

A solution to preserve compatibility with the previous `nginx.conf` files could be to check if there is a `server{}` in the file, if so, we would ignore the template file and do the `sed` job on that file.

smashwilson commented:
This is very cool. :+1: I've left a bit of feedback on the PR, if you don't mind :bow:

jbtrystram commented:
Thanks you !!  :smile: your comments are very helpful ! I absolutely don't mind, comments and mutual improvement is the whole point of open-source and PRs, isn'it ?
I am quite new in contributing in open source projects, so i hadn't realize i should do multiples PRs and more commits.
Should I update this PR and break it down into several ones ? Or leave that and add commits when we have found a solution to the issues you raised above?

rgbkrk commented:
My main suggestion about breaking it up is that I could tell that some parts of this would get merged immediately while others would go back and forth. I didn't want us to hold back the easy ones for the others.

Thank you for contributing!

jbtrystram committed:
Update let's encrypt command

Since the `--authenticator` option is depreciated.

jbtrystram committed:
updated alpine to the latest version

jbtrystram commented:
So there is a design choice to be done here, about the templating system : 
There are two different options i can think about : 
- Ignore the default-template-file if a `server {}` section exists in the main `nginx.conf` file, and if more than 1 domain is passed in the `DOMAIN` env variable, then the `server{}` section is duplicated for each domain.
- Allow a specific vhost file to created in the templates folder, eg : `templates/domain1.com.conf`, if so use it, if not use the default one (or the `server{}` directive in the main `nginx.conf` file)

I think that both options can be done in the same behaviour, actually.

smashwilson commented:
That's sounding fairly complex. What I had in mind was something much simpler:
1. Run the `sed` calls on `templates/nginx.conf` with the raw values of `${DOMAIN}` and `${UPSTREAM}` as-is, regardless of its content. If the template has been left as the default, this will do nothing, because it won't contain any occurrences of either `$DOMAIN` or `$UPSTREAM`. But, if a user is running a derived image with a custom `templates/nginx.conf`, this lets their image continue to work even after this is merged.
2. Use the `for` loop over `${DOMAINSARRAY}` to render `/templates/vhost.sample.conf` with each successive DOMAIN value, as you've already done, to create more `/etc/nginx/vhosts/` entries.

Step 1 will have no effect on the default `templates/nginx.conf`. Likewise, step 2 will have no effect on a customized `templates/nginx.conf` that doesn't use an `include` directive.

Mostly I'm concerned about not breaking things for people who are already using this behavior! At least @itzg is already taking advantage of the ability to template `nginx.conf`, and we don't (yet) have any tags to anchor to, so we need to be extra careful about suddenly breaking existing images that use `smashwilson/lets-nginx:latest`.

itzg commented:
Thanks for pinging me @smashwilson . Yes, I am actively using the ability to customize the nginx config via a derivative image; however, I can adapt to any design changes here.

Also, if it helps, in my particular use case I only need to adjust the `client_max_body_size` in the outer `http` section ([here](https://github.com/moorkop/mccy-lets-nginx/blob/master/nginx.conf#L20)). I didn't have a need to customize the `server` section, therefore per-domain adjustments aren't needed. Again...that's just my use case :)

Looking at the changes, since `bash` is now being used I would recommend switching the templating to some [bash native tricks like this.](http://pempek.net/articles/2013/07/08/bash-sh-as-template-engine/). I had used `sed` only as an effort to stick with the smaller footprint of `sh`

jbtrystram committed:
Ask only one certificate to letsencrypt

to avoid hiting the 5-certs limit.
The let's encrypt comand passes a `--domain yourdomain.tld` for each domain provided in `$DOMAIN`

jbtrystram committed:
refactored the for loop

jbtrystram committed:
Process nginx.conf for backward compatibility

The raw values of `$DOMAIN` and `$UPSTREAM` to process nginx.conf and keep backward compatibility as suggested by @smashwilson here : https://github.com/smashwilson/lets-nginx/pull/9#issuecomment-199817032

jbtrystram committed:
typo fix in README

smashwilson commented on entrypoint.sh:
```diff
@@ -41,41 +57,58 @@ chown nginx:nginx /var/cache/nginx
+mkdir -p /etc/nginx/vhosts/
+
+# Process the nginx.conf with ra values of $DOMAIN and $UPSTREAM to ensure backward-compatibility
+  dest="/etc/nginx/vhosts/nginx.conf"
```
This should be `dest="/etc/nginx/nginx.conf"`, right?

jbtrystram replied:
oopsie ! Good catch !

jbtrystram committed:
README update

jbtrystram committed:
error in the destination of nginx.conf

jbtrystram committed:
Merge branch 'letsencrypt' into multiples_domains

jbtrystram commented:
Voilà ! :)

smashwilson commented:
Excellent. Thank you for taking this on and for your patience!
